web/src/lib/components/pages.tsx

```diff
@@ -237,47 +237,29 @@ export default class Pages extends Component<PagesProps, PagesState> {
     });
   }
 
-  private uploadRecordings(recordings: any[],
+  private async uploadRecordings(recordings: any[],
                            sentences: string[],
                            progressCb: Function): Promise<void> {
-
-    return new Promise<void>((res, rej) => {
-      this.ensurePrivacyAgreement().then(() => {
-        let runningTotal = 0;
-        let originalTotal = recordings.length;
-
-        // This function calls itself recursively until
-        // all recordings are uploaded.
-        let uploadNext = () => {
-          if (recordings.length === 0) {
-            this.props.api.uploadDemographicInfo().then(() => {
-              return;
-            });
-            res();
-            return;
-          }
-
-          let recording = recordings.pop();
-          let blob = recording.blob;
-          let sentence = sentences.pop();
-
-          this.props.api.uploadAudio(blob, sentence).then(() => {
-            runningTotal++;
-            let percentage = Math.floor((runningTotal / originalTotal) * 100);
-            progressCb && progressCb(percentage);
-            this.props.user.tallyRecording();
-            uploadNext();
-          });
-        };
-
-        // Start the recursive chain to upload the recordings serially.
-        uploadNext();
-      }).catch(rej);
-    }).then(() => {
+    try {
+      await this.ensurePrivacyAgreement();
+      const originalTotal = recordings.length;
+
+      for (let runningTotal = 1; runningTotal <= originalTotal; runningTotal++) {
+        const recording = recordings.pop();
+        const blob = recording.blob;
+        const sentence = sentences.pop();
+
+        await this.props.api.uploadAudio(blob, sentence);
```

Comment:
note: if this fails, i believe it will now forward that error to the caller of this function, which didn't happen in the previous Promise structure. also note, your way is better. we probably want those errors :)